src/state/safety.rs

```diff
@@ -0,0 +1,315 @@
+/// to have the view of its prepare qc set as the locked view.
+///
+/// For this, it is necessary that:
+/// 1. Its chain ID matches the chain ID of the replica, or is the genesis qc.
+/// 2. It justifies a known block, or is the genesis qc.
+/// 3. Its view number is greater than the locked qc view or its block extends from the locked block.
+/// 4. If it is a prepare, precommit, or commit qc, the block it justifies has pending validator state
+///    updates.
+/// 5. If its qc is a generic qc, the block it justifies *does not* have pending validator set updates.
+///
+/// # Precondition
+/// [QuorumCertificate::is_correct] holds for block.justify.
+pub fn safe_qc<K: KVStore>(qc: &QuorumCertificate, block_tree: &BlockTree<K>, chain_id: ChainID) -> Result<bool, BlockTreeError> {
+    Ok(
+        /* 1 */ (qc.chain_id == chain_id || qc.is_genesis_qc()) &&
+        /* 2 */ (block_tree.contains(&qc.block) || qc.is_genesis_qc()) &&
+        /* 3 */ (qc.view > block_tree.locked_qc()?.view || extends_locked_qc_block(qc, block_tree)?) &&
+        /* 4 */ (((qc.phase.is_prepare() || qc.phase.is_precommit() || qc.phase.is_commit() || qc.phase.is_decide()) && 
+                   block_tree.validator_set_updates_status(&qc.block)?.contains_updates()) ||
+        /* 5 */ (qc.phase.is_generic() && !block_tree.validator_set_updates_status(&qc.block)?.contains_updates()))
+    )
+}
+
+/// Returns whether a nudge can be considered safe, and hence cause updates to the block tree.
+/// For this, it is necessary that:
+/// 1. safe_qc(&nudge.justify, block_tree, chain_id).
+/// 2. nudge.justify is a prepare, precommit, or commit qc.
+/// 3. Its chain ID matches the chain ID of the replica.
+/// 4. &nudge.justify is either a commit qc, or &nudge.justify.view = cur_view - 1.
+/// 
+/// This method enforces the commit rule for validator-set-updating blocks.
+/// 
+/// # Precondition
+/// [QuorumCertificate::is_correct] holds for nudge.justify.
+pub fn safe_nudge<K: KVStore>(nudge: &Nudge, cur_view: ViewNumber, block_tree: &BlockTree<K>, chain_id: ChainID) -> Result<bool, BlockTreeError> {
+    Ok(
+        /* 1 */ safe_qc(&nudge.justify, block_tree, chain_id)? &&
+        /* 2 */ nudge.justify.is_nudge_justify() &&
+        /* 3 */ (nudge.chain_id == chain_id) &&
+        /* 4 */ (nudge.justify.phase.is_commit() || nudge.justify.view == cur_view - 1)
+    )
+}
+
+/// Returns an optional qc that should be set as the new locked qc after seeing a given justify. This
+/// should be called whenever a new, correct and safe justify qc is seen, whether through a block
+/// proposal or through a nudge. This method implements the following rule:
+/// 1. If a block justify is seen,
+///     - if it is a decide qc, then locked qc should be updated to the block justify.
+///     - else the locked qc should be updated to the parent qc of the block justify.
+/// 2. If a nudge justify is seen,
+///     - if it is a precommit qc, then locked qc should be updated to the precommit qc.
```

Comment:
Why is it that when we receive a Generic QC, the Locked QC should be updated to the Justify of `locked_qc.block`, whereas when we receive a Decide QC or a Precommit QC, the Justify should be updated to the received QC itself?

I'm guessing this is a choice based on convenience as opposed to a choice that is significant to the safety and liveness of the protocol, correct? That is, when we receive a Decide QC, we don't have access to the Commit QC, and when we receive a Precommit QC, we don't have access to the Prepare QC, so we can't lock the Commit QC and Prepare QC respectively and we can only lock the receive QC itself.

Reply:
I thought about this choice for a couple hours and my quick assessment is that this indeed does not do anything that weakens or threatens liveness. The reason why hinges on the fact that we check for "consecutive views" when deciding whether or not to vote for a Nudge.

While thinking about this, I also started thinking about a related design choice that we've made, namely, to have "1 phase = 1 view" for the phased protocol, instead of "3 phases = 1 view" like in phased original HotStuff.

I feel like if we had "3 phases = 1 view" we wouldn't even need to enforce "consecutive views" for the phased protocol. I wonder whether there is an explicit reason behind why we chose "1 phase = 1 view" over "3 phases = 1 view" besides that fact that this is what is done in HotStuff-rs v0.2 and v0.3?

Reply:
First, when we receive a generic QC, call it `qc`, then lockedQC is updated to the the justify of `qc.block`, not justify of `locked_qc.block`. This is the pipelined HotStuff way of locking. We lock a prepare QC for the block when there is a precommitQC, but in pipelined hotstuff this means we lock the justify of of `qc.block` (prepareQC for the parent of `qc.block`) when we see `qc` (precommitQC for the parent of `qc.block`) . So this is correct.

As for locking on precommit QC this is a correct simplification/adaptation of the protocol to non-pipelined HotStuff for validator set updating blocks. Since the views must be consecutive for prepare, precommit, and commit QCs, to avoid storing the prepareQC we can just lock the precommitQC on seeing it. Effectively, we are locking the block, not the QC - the QC is just there to store the view in which the validators voted for the block, so that in case we see that the quorum has voted for a conflicting block in a latter view we drop the lock (important for liveness!). Since the views are consecutive, it doesn't matter which view we store since there can be no QC with a view between them.

As for locking on seeing decideQC, this is an adaptation for sync, since when in sync the sync client does not get to see the precommitQC for a vs-updating block. But since it's possible that a vs-updating block B is at the head of the blockchain when the replica catches up with others via sync, it needs to lock B since B was committed. Again recall that it's really about locking blocks not QCs, and a block that is committed must be "locked" in the sense that the `safe_qc` predicate cannot hold for it.